packages/cohort/src/replicator2/cohort_suffix.rs

```diff
@@ -0,0 +1,368 @@
+        self.messages[index] = Some(new_sfx_item);
+        Ok(())
+    }
+
+    pub fn set_decision_outcome(&mut self, version: u64, decision_outcome: Decision) {
+        if version >= self.meta.head {
+            let index = self.index_from_head(version).unwrap();
+            if let Some(Some(item_to_update)) = self.messages.get_mut(index) {
+                item_to_update.item.decision = Some(decision_outcome);
+            } else {
+                log::warn!(
+                    "Unable to update decision as message with version={version} not found. len: {}, index: {}, head: {}",
+                    self.messages.len(),
+                    index,
+                    self.meta.head
+                );
+            }
+        }
+    }
+
+    pub fn set_safepoint(&mut self, version: u64, safepoint: Option<u64>) {
+        if version >= self.meta.head {
+            let index = self.index_from_head(version).unwrap();
+            if let Some(Some(item_to_update)) = self.messages.get_mut(index) {
+                item_to_update.item.safepoint = safepoint;
+            } else {
+                log::warn!("Unable to update safepoint as message with version={version} not found");
+            }
+        }
+    }
+
+    pub fn set_item_installed(&mut self, version: u64) {
+        if version >= self.meta.head {
+            let index = self.index_from_head(version).unwrap();
+            if let Some(Some(item_to_update)) = self.messages.get_mut(index) {
+                item_to_update.is_installed = true;
+                item_to_update.is_in_flight = false;
+            } else {
+                log::warn!("Unable to update is_installed flag as message with version={version} not found");
+            }
+        }
+    }
+
+    pub fn set_item_in_flight(&mut self, version: u64) {
+        if version >= self.meta.head {
+            let index = self.index_from_head(version).unwrap();
+            if let Some(Some(item_to_update)) = self.messages.get_mut(index) {
+                item_to_update.is_in_flight = true;
+            } else {
+                log::warn!("Unable to update is_in_flight flag as message with version={version} not found");
+            }
+        }
+    }
+
+    pub fn find_new_decided(&mut self, last_in_flight: Option<u64>, skip: bool) -> Option<SuffixItem> {
+        let skip_count = if skip {
+            if let Some(last_in_flight) = last_in_flight {
+                match self.index_from_head(last_in_flight) {
+                    None => 0,
+                    Some(index) => index + 1,
+                }
+            } else {
+                0
+            }
+        } else {
+            0
+        };
+
+        self.messages
+            .iter()
+            .skip(skip_count)
+            .flatten()
+            .take_while(|i| i.is_decided)
+            .filter(|i| !i.is_installed && !i.is_in_flight)
+            .take(1)
```

Comment:
This could be reason why the performance is lower when I compared the performance between the way `replicator2` creates statemap vs the changes I made in this [branch](https://github.com/kindredgroup/talos/blob/feature/replictor-perf-updates/packages/cohort/src/replicator/suffix.rs#L97).

We do all the iteration, flatten and filter and take only 1 out of it. And this function is called in a loop.

In the branch I shared above, you can see, we pick all the eligible records, and later there is a loop to further narrow down  and create the statemap.

Reply:
GK is raising JIRA ticket for this item based on new temp branch